pkg/xds/sync/proxy_builder_test.go

```diff
@@ -0,0 +1,158 @@
+		bytes := []byte(rawResource)
+
+		res, err := rest_types.UnmarshallToCore(bytes)
+		Expect(err).To(BeNil())
+
+		switch res.Descriptor().Name {
+		case core_mesh.ZoneEgressType:
+			zoneEgress := res.(*core_mesh.ZoneEgressResource)
+			Expect(resourceManager.Create(ctx, zoneEgress, store.CreateBy(core_model.MetaToResourceKey(zoneEgress.GetMeta())))).To(Succeed())
+		case core_mesh.ZoneIngressType:
+			zoneIngress := res.(*core_mesh.ZoneIngressResource)
+			Expect(resourceManager.Create(ctx, zoneIngress, store.CreateBy(core_model.MetaToResourceKey(zoneIngress.GetMeta())))).To(Succeed())
+		case core_mesh.DataplaneType:
+			dataplane := res.(*core_mesh.DataplaneResource)
+			Expect(resourceManager.Create(ctx, dataplane, store.CreateBy(core_model.MetaToResourceKey(dataplane.GetMeta())))).To(Succeed())
+		case core_mesh.MeshType:
+			meshResource := res.(*core_mesh.MeshResource)
+			Expect(resourceManager.Create(ctx, meshResource, store.CreateBy(core_model.MetaToResourceKey(meshResource.GetMeta())))).To(Succeed())
+		case core_mesh.ExternalServiceType:
+			externalService := res.(*core_mesh.ExternalServiceResource)
+			Expect(resourceManager.Create(ctx, externalService, store.CreateBy(core_model.MetaToResourceKey(externalService.GetMeta())))).To(Succeed())
+		}
+	}
+}
+
+var _ = Describe("Proxy Builder", func() {
+	core.Now = func() time.Time {
+		now, _ := time.Parse(time.RFC3339, "2018-07-17T16:05:36.995+00:00")
+		return now
+	}
+	tracker := mockMetadataTracker{}
+	localZone := "zone-1"
+
+	ctx := context.Background()
+	config := kuma_cp.DefaultConfig()
+	config.Multizone.Zone.Name = localZone
+	builder, err := test_runtime.BuilderFor(ctx, config)
+	Expect(err).ToNot(HaveOccurred())
+	rt, err := builder.Build()
+	Expect(err).ToNot(HaveOccurred())
+
+	meshCtxBuilder := xds_context.NewMeshContextBuilder(
+		rt.ReadOnlyResourceManager(),
+		server.MeshResourceTypes(server.HashMeshExcludedResources),
+		rt.LookupIP(),
+		rt.Config().Multizone.Zone.Name,
+		vips.NewPersistence(rt.ReadOnlyResourceManager(), rt.ConfigManager()),
+		rt.Config().DNSServer.Domain,
+	)
+	metrics, err := core_metrics.NewMetrics("cache")
+	Expect(err).ToNot(HaveOccurred())
+	meshCache, err := mesh.NewCache(rt.Config().Store.Cache.ExpirationTime, meshCtxBuilder, metrics)
+	Expect(err).ToNot(HaveOccurred())
+	initializeStore(ctx, rt.ResourceManager(), "default_resources.yaml")
+
+	Describe("Build() zone egress", func() {
+		egressProxyBuilder := sync.DefaultEgressProxyBuilder(
+			ctx,
+			rt,
+			tracker,
+			meshCache,
+			envoy_common.APIV3,
+		)
+
+		It("should build proxy object for egress", func() {
+			// given
+			rk := core_model.ResourceKey{Name: "zone-egress-1"}
+
+			// when
+			proxy, err := egressProxyBuilder.Build(rk)
+			Expect(err).ToNot(HaveOccurred())
+			proxyYaml, err := ToYAML(proxy)
+			Expect(err).ToNot(HaveOccurred())
+
+			// then
+			Expect(proxyYaml).To(matchers.MatchGoldenYAML(filepath.Join("testdata", "01.zone-egress-proxy.golden.yaml")))
```

Comment:
This comment refers also to all the existing tests that do this, of which I know there are many, so I know it's the existing style. Not picking on this PR! Just trying to stop any new ones from being introduced.

I don't think we should have to mock `Now` in order to write tests. And the tests shouldn't test YAML equality. YAML is not the interface of this module. For e2e, ok, YAML is the e2e interface. But here we should assert on the parts of the objects that matter for what we're testing, which would spare us having to mock `Now`. The tests can be rewritten, see [#2902](https://github.com/kumahq/kuma/pull/2902/files#diff-7636744e66c60614da0142f1be5970020ab9ad7f28c373037e1d9082015c6344)

Reply:
I can agree that YAML comparison isn't the best for that type of test. My concern is that to cover the same number of objects that are generated by the method would require hundreds of lines of code just for the expected format definition. Also, I’ve tried to compare specific objects that I’ve expected to generate but here I want to know more than just version change(Correct IP change/available services/outbound map). Also, the returned object had metadata fields defined with random fields.

Reply:
I think a lot of code is being tested here which isn't part of this module.

```
		meshResources := &xds.MeshResources{
			Mesh:             mesh,
			TrafficRoutes:    trafficRoutes,
			ExternalServices: externalServices,
			EndpointMap: xds_topology.BuildRemoteEndpointMap(
				mesh,
				p.zone,
				zoneIngresses,
				externalServices,
				meshCtx.DataSourceLoader,
			),
			ExternalServicePermissionMap: permissions.BuildExternalServicesPermissionsMapForZoneEgress(
				externalServices,
				trafficPermissions,
			),
			ExternalServiceFaultInjections: faultinjections.BuildExternalServiceFaultInjectionMapForZoneEgress(
				externalServices,
				faultInjections,
			),
			ExternalServiceRateLimits: ratelimits.BuildExternalServiceRateLimitMapForZoneEgress(
				externalServices,
				rateLimits,
			),
		}
```

Testing all of that seems like just testing `xds_topology`, `permissions`, `faultinjections` and `ratelimits` again.

Can the test be made to cover the logic in this module? For example maybe the test can check whether a specific fault injection is or isn't present depending on the resource key given to `Build`? I.e. that we _call_ those modules correctly.

This way it's also clearer what this test is actually testing and it makes it less fragile if those other modules change their output.

Reply:
Removed golden file and changed the test to compare only part of the data.

Reply:
Nice!

Reply:
This is an interesting conversation that should probably be taken to a wider audience.